### PR TITLE
Fix subject reduction VS cumulative inductives and function eta

### LINKED ISSUE #7015: Polymorphic cumulativity and function eta break subject reduction
Present in 8.7 and current master.
```coq
Set Universe Polymorphism.
Set Polymorphic Inductive Cumulativity.
Set Printing Universes.

(* in the real world foo@{i} might be [@paths@{i} nat] I guess *)
Inductive foo : nat -> Type :=.

(* on [refl (fun x => f x)] this computes to [refl f] *)
Definition eta_out {A B} (f g : forall x : A, B x) (e : (fun x => f x) = (fun x => g x)) : f = g.
Proof.
  change (f = g) in e. destruct e;reflexivity.
Defined.

Section univs.
  Universes i j.
  Constraint i < j. (* constraint not needed in master *)

  Definition one : (fun n => foo@{i} n) = fun n => foo@{j} n := eq_refl.

  Definition two : foo@{i} = foo@{j} := eta_out _ _ one.

  Fail Definition two' : foo@{i} = foo@{j} := Eval compute in two.

  (* Failure of SR doesn't just mean that the type changes, sometimes
     we lose being well-typed entirely. *)
  Definition three := @eq_refl (foo@{i} = foo@{j}) two.
  Fail Definition four := Eval compute in three.
End univs.
```

### PR DESCRIPTION
**Pull request by SkySkimmer**
Fix #7015

### CODE REVIEW

**Comment by SkySkimmer** [comment]
Bench probably won't say much as the only path changed is when using cumulative inductives but may still be worth running.

File: kernel/reduction.ml [hunk 1]
```diff
@@ -336,6 +339,35 @@ let is_irrelevant infos lft c =
...
   try Relevanceops.relevance_of_fterm env (info_relevances infos.cnv_inf) lft c == Sorts.Irrelevant with _ -> false
 
+let identity_of_ctx (ctx:Constr.rel_context) =
+  let rec aux acc i = function
+    | [] -> Array.rev_of_list acc
+    | Context.Rel.Declaration.LocalDef _ :: ctx -> aux acc (i+1) ctx
+    | Context.Rel.Declaration.LocalAssum _ :: ctx -> aux (mkRel i :: acc) (i+1) ctx
+  in
+  aux [] 1 ctx
+
```
Comment (mattam82) [inline kernel/reduction.ml:349]: That's a variant of `Context.Rel.to_extended_vect`

**Comment by mattam82** [comment]
This looks right, dynamically eta-expanding constructor and inductive applications so that we can properly test cumulativity. This will probably need to be reflected in evar-conv/w_unify too...

**Comment by mattam82** [comment]
It would be worth checking in the initial example which had a strong elimination on a boolean two.

**Comment by SkySkimmer** [comment]
updated

**Review by ppedrot — COMMENTED** [review]
This looks intuitively correct but I wouldn't mind a little session around a whiteboard with @SkySkimmer and @mattam82 to fully convince myself.

**Comment by SkySkimmer** [comment]
~~~

┌─────────────────────────────┬─────────────────────────┬─────────────────────────────────────────────┬─────────────────────────────────────────────┬───────────────────────────────┬───────────────────┐
│                             │      user time [s]      │                 CPU cycles                  │              CPU instructions               │     max resident mem [KB]     │    mem faults     │
│                             │                         │                                             │                                             │                               │                   │
│                package_name │     NEW     OLD PDIFF   │               NEW               OLD PDIFF   │               NEW               OLD PDIFF   │        NEW        OLD PDIFF   │  NEW  OLD PDIFF   │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│            coq-engine-bench │ 1292.24 1300.21 -0.61 % │     3620702027817     3641302053969 -0.57 % │     6256513799735     6261015934811 -0.07 % │    4299112    4299124 -0.00 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│                coq-bedrock2 │  227.78  228.70 -0.40 % │      630592605291      634352893417 -0.59 % │     1050827373438     1050675754751 +0.01 % │    1087232    1087224 +0.00 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│              coq-coquelicot │   73.54   73.83 -0.39 % │      202355404556      202541484708 -0.09 % │      263914904904      263931191087 -0.01 % │     607344     607344 +0.00 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│       coq-mathcomp-fingroup │   49.86   50.05 -0.38 % │      138027440828      138365200718 -0.24 % │      191218581161      191220472936 -0.00 % │     485816     486300 -0.10 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│       coq-performance-tests │ 2868.22 2877.99 -0.34 % │     7998404106029     8027811484389 -0.37 % │    15000881438705    14998959563369 +0.01 % │    4316480    4316592 -0.00 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│             coq-fiat-crypto │ 5284.82 5302.12 -0.33 % │    14683136608913    14733412582868 -0.34 % │    25254790076620    25251039331611 +0.01 % │    2422144    2407132 +0.62 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│                 coq-bignums │   60.20   60.38 -0.30 % │      166318107023      166978174261 -0.40 % │      222598421242      222593128153 +0.00 % │     474576     474648 -0.02 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│                coq-rewriter │  630.51  632.38 -0.30 % │     1749210335485     1752630494570 -0.20 % │     2670970956115     2671050023035 -0.00 % │     991448     991548 -0.01 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│               coq-fourcolor │ 2509.20 2515.95 -0.27 % │     6986762673358     7014443837924 -0.39 % │    11480305603835    11480222284208 +0.00 % │     764280     764192 +0.01 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│      coq-mathcomp-odd-order │ 1130.24 1132.98 -0.24 % │     3148407555489     3154914133641 -0.21 % │     5095692961768     5095314189447 +0.01 % │    1039764    1039820 -0.01 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│                    coq-corn │ 1772.20 1776.18 -0.22 % │     4938036576709     4946449514784 -0.17 % │     7601806002857     7601733526616 +0.00 % │     843988     843832 +0.02 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│                 coq-unimath │ 4064.43 4073.53 -0.22 % │    11303262850203    11333448929551 -0.27 % │    21563034202161    21558326640066 +0.02 % │    1007832    1007744 +0.01 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│                coq-coqprime │   85.46   85.60 -0.16 % │      234628781056      235312504459 -0.29 % │      332724126321      332727649767 -0.00 % │     770040     769720 +0.04 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│      coq-mathcomp-character │  163.97  164.14 -0.10 % │      456276244820      456344044375 -0.01 % │      634147547533      634088121136 +0.01 % │     727384     727456 -0.01 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│          coq-mathcomp-field │  295.25  295.51 -0.09 % │      822533669232      823326374385 -0.10 % │     1269676130288     1269553564361 +0.01 % │     815488     815640 -0.02 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│                    coq-hott │  343.82  344.05 -0.07 % │      933876980824      933148754400 +0.08 % │     1509655253130     1509400520435 +0.02 % │     621744     648272 -4.09 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│       coq-mathcomp-solvable │  175.95  176.02 -0.04 % │      490465731208      489924870474 +0.11 % │      693551066870      693449495320 +0.01 % │     719548     735388 -2.15 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│ coq-rewriter-perf-SuperFast │ 1038.82 1039.18 -0.03 % │     2890726374643     2891582539226 -0.03 % │     4609267071203     4610101508837 -0.02 % │    1071076    1071052 +0.00 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│             coq-lambda-rust │ 1539.58 1538.86 +0.05 % │     4288079679400     4285628621389 +0.06 % │     6113301781397     6115355943964 -0.03 % │    1214716    1213776 +0.08 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│            coq-fiat-parsers │  670.87  670.45 +0.06 % │     1878368081438     1878237492585 +0.01 % │     2836560488807     2836520121549 +0.00 % │    3660084    3660276 -0.01 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│              coq-verdi-raft │ 1279.28 1278.12 +0.09 % │     3564543824977     3559592178948 +0.14 % │     4981525263251     4981340143802 +0.00 % │     978692     978608 +0.01 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│                         coq │  567.58  567.01 +0.10 % │     1588141761673     1588723111788 -0.04 % │     2162383711304     2162404699267 -0.00 % │     602812     602608 +0.03 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│        coq-mathcomp-algebra │  143.56  143.38 +0.13 % │      399246467162      398821294963 +0.11 % │      518897431035      518943587610 -0.01 % │     562264     562256 +0.00 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│                  coq-geocoq │ 1512.39 1509.74 +0.18 % │     4215468690910     4208620512032 +0.16 % │     6401133026246     6401033932354 +0.00 % │    1211560    1211376 +0.02 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│                   coq-flocq │  173.05  172.61 +0.25 % │      480746665613      479640985922 +0.23 % │      604307529217      604332474194 -0.00 % │     900932     900908 +0.00 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│                   coq-verdi │  122.59  122.21 +0.31 % │      338931892294      338118321973 +0.24 % │      422767617409      422693636305 +0.02 % │     563712     563716 -0.00 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│                coq-compcert │  700.16  697.83 +0.33 % │     1948200862529     1945670626007 +0.13 % │     2705226221438     2705104896403 +0.00 % │    1134692    1134696 -0.00 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│                   coq-color │  690.67  688.10 +0.37 % │     1925836608511     1918536512748 +0.38 % │     2457759278753     2457683598030 +0.00 % │    1491588    1491700 -0.01 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│      coq-mathcomp-ssreflect │   52.92   52.71 +0.40 % │      144761503855      144220425131 +0.38 % │      186873705020      186891896860 -0.01 % │     516588     516592 -0.00 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│                 coq-coqutil │   62.00   61.71 +0.47 % │      171932162569      170930908179 +0.59 % │      215426259953      215425671997 +0.00 % │     534460     534548 -0.02 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│               coq-fiat-core │  111.88  111.26 +0.56 % │      318940790050      317422513327 +0.48 % │      426144374095      426116119757 +0.01 % │     494032     494212 -0.04 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│            coq-math-classes │  214.30  212.94 +0.64 % │      598464186144      595870499814 +0.44 % │      804908642335      804906287041 +0.00 % │     513556     513560 -0.00 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│                  coq-sf-plf │   45.20   44.74 +1.03 % │      125520643530      124928632733 +0.47 % │      171897305637      171858552818 +0.02 % │     506904     514624 -1.50 % │    0    0  +nan % │
└─────────────────────────────┴─────────────────────────┴─────────────────────────────────────────────┴─────────────────────────────────────────────┴───────────────────────────────┴───────────────────┘
~~~

**Comment by mattam82** [comment]
@ppedrot we can do that. Note that this relies on domains of lambdas to be properly compared during conversion. We would have to do a comparison of contexts instead if we wanted the algorithm to maintain the invariant that the two terms have a common supertype.

**Comment by mattam82** [comment]
We discussed again with Gaëtan and did not find any issue with this formulation (comparing contexts would be more heavy  with the current APIs), so I'm merging this.

**Review by mattam82 — APPROVED** [review]

**Comment by mattam82** [comment]
@coqbot: merge now

**Pull request merged**